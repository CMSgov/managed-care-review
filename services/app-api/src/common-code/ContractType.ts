import type {
    Contract,
    ContractPackageSubmission,
    ContractRevision,
} from '../gen/gqlServer'

const getLastContractSubmission = (
    contract: Contract
): ContractPackageSubmission | undefined => {
    return (
        (contract.packageSubmissions && contract.packageSubmissions[0]) ??
        undefined
    )
}

const getContractRev = (contract: Contract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY'
}

const isBaseContract = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract): boolean =>
    isContractAmendment(contract) ||
    (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
