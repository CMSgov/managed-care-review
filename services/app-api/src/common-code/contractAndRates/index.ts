export {
    isBaseContract,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isContractAmendment,
    isSubmitted,
    isContractWithProvisions,
    getLastContractSubmission,
} from './contractData'

export {
    getProvisionDictionary,
    sortModifiedProvisions,
    generateApplicableProvisionsList,
    generateProvisionLabel,
    isMissingProvisions,
    hasValidModifiedProvisions,
} from './ContractProvisions'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
    isMedicaidBaseProvision,
} from './ModifiedProvisions'
