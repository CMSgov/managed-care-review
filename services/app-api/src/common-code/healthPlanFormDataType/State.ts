type StateType = {
    code: string
    name: string
    programs: ProgramArgType[]
}

interface ProgramArgType {
    id: string
    name: string // short name. This is used most often in the application including in submission name
    fullName: string // full name is used in submission summary page
    isRateProgram: boolean // specifies if program relates to rates rather than contract
}

export type { StateType, ProgramArgType }
