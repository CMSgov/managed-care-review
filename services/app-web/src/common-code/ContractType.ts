import { Contract, ContractRevision, UnlockedContract } from '../gen/gqlClient'
import { getLastContractSubmission } from '../gqlHelpers/contractsAndRates'

const getContractRev = (contract: Contract | UnlockedContract): ContractRevision | undefined => {
    if (contract.draftRevision) {
        return contract.draftRevision
    } else {
        return getLastContractSubmission(contract)?.contractRevision
    }
}
const isContractOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_ONLY' 
}


const isBaseContract = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'BASE'
}

const isContractAmendment = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.contractType === 'AMENDMENT'
}

const isCHIPOnly = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.populationCovered === 'CHIP'
}

const isContractAndRates = (contract: Contract | UnlockedContract): boolean => {
    const contractRev = getContractRev(contract)
    return contractRev?.formData?.submissionType === 'CONTRACT_AND_RATES'
}

const isContractWithProvisions = (contract: Contract | UnlockedContract): boolean =>
    isContractAmendment(contract) || (isBaseContract(contract) && !isCHIPOnly(contract))

const isSubmitted = (contract: Contract | UnlockedContract): boolean =>
    contract.status === 'SUBMITTED'

export {
    isContractWithProvisions,
    isBaseContract,
    isContractAmendment,
    isCHIPOnly,
    isContractOnly,
    isContractAndRates,
    isSubmitted,
}
