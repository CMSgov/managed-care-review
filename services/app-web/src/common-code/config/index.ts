export { assertIsAuthMode, assertNever } from './config'
export type { AuthModeType } from './config'
