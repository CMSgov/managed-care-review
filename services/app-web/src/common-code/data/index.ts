export {removeNullAndUndefined, replaceNullsWithUndefineds } from './dataUtilities'