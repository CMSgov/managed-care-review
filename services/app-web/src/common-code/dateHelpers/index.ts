export { formatCalendarDate, formatRateNameDate } from './calendarDate'
export { formatGQLDate } from './gqlDate'
export { dayjs } from './dayjs'
