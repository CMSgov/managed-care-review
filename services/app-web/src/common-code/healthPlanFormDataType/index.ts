// TEMPORARY: these files are embedded in app-web for now b/c
// CRA prevents you from importing code outside of /src
// The fix is to use yarn workspaces to allow us to import shared packages
// Domain Types
export type {
    ActuarialFirmType,
    ActuaryCommunicationType,
    ActuaryContact,
    ContractAmendmentInfo,
    ContractExecutionStatus,
    ContractType,
    DocumentCategoryType,
    UnlockedHealthPlanFormDataType,
    ManagedCareEntity,
    RateType,
    RateCapitationType,
    StateContact,
    SubmissionDocument,
    SubmissionType,
    RateInfoType,
    PopulationCoveredType,
    SharedRateCertDisplay,
} from './UnlockedHealthPlanFormDataType'

export type {
    FederalAuthority,
    CHIPFederalAuthority,
} from './FederalAuthorities'

export type {
    ModifiedProvisionsMedicaidAmendment,
    ModifiedProvisionsMedicaidBase,
    CHIPModifiedProvisions,
    GeneralizedProvisionType,
    MedicaidAmendmentProvisionType,
    GeneralizedModifiedProvisions,
} from './ModifiedProvisions'

export {
    federalAuthorityKeys,
    federalAuthorityKeysForCHIP,
} from './FederalAuthorities'

export {
    modifiedProvisionMedicaidBaseKeys,
    modifiedProvisionMedicaidAmendmentKeys,
    provisionCHIPKeys,
    isCHIPProvision,
    isMedicaidAmendmentProvision,
} from './ModifiedProvisions'

export type { LockedHealthPlanFormDataType } from './LockedHealthPlanFormDataType'
export {
    hasValidContract,
    hasValidDocuments,
    hasValidRates,
    hasAnyValidRateData,
    hasValidSupportingDocumentCategories,
    isContractAndRates,
    isContractOnly,
    isUnlockedHealthPlanFormData,
    isLockedHealthPlanFormData,
    isSubmitted,
    programNames,
    packageName,
    generateRateName,
    convertRateSupportingDocs,
    removeInvalidProvisionsAndAuthorities,
    removeRatesData,
    hasValidPopulationCoverage,
    isValidAndCurrentLockedHealthPlanFormData,
} from './healthPlanFormData'

export type { HealthPlanFormDataType } from './HealthPlanFormDataType'
export type { ProgramArgType, StateType } from './State'

export { StateCodes, isValidStateCode } from './StateCodeType'
export type { StateCodeType } from './StateCodeType'
