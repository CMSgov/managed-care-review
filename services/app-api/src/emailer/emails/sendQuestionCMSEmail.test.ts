import {
    testEmailConfig,
    mockContractRev,
    mockMNState,
} from '../../testHelpers/emailerHelpers'
import type { CMSUserType, StateType, Question } from '../../domain-models'
import { packageName } from 'app-web/src/common-code/healthPlanFormDataType'
import { sendQuestionCMSEmail } from './index'

const stateAnalysts = [
    'stateAnalysts1@example.com',
    'stateAnalysts1@example.com',
]

const flState: StateType = {
    stateCode: 'FL',
    name: 'Florida',
}

const cmsUser: CMSUserType = {
    id: '1234',
    role: 'CMS_USER',
    divisionAssignment: 'DMCO',
    familyName: 'McDonald',
    givenName: 'Ronald',
    email: 'cms@email.com',
    stateAssignments: [flState],
}

const question: Question = {
    id: '1234',
    contractID: 'contract-id-test',
    createdAt: new Date('01/01/2024'),
    addedBy: cmsUser,
    documents: [],
    division: 'DMCO',
    responses: [],
}

test('to addresses list only includes state analyst when a DMCO user submits a question', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs

    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        question
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.not.objectContaining({
            toAddresses: expect.arrayContaining([
                ...testEmailConfig().oactEmails,
                ...testEmailConfig().dmcpEmails,
            ]),
        })
    )

    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining([...stateAnalysts]),
        })
    )
})

test('to addresses list includes state analyst and OACT group emails when an OACT user submits a question', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const oactUser: CMSUserType = {
        ...cmsUser,
        divisionAssignment: 'OACT',
    }
    const questionFromOACT: Question = {
        ...question,
        addedBy: oactUser,
    }
    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        questionFromOACT
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining([
                ...stateAnalysts,
                ...testEmailConfig().oactEmails,
            ]),
        })
    )
})

test('to addresses list includes state analyst and DMCP group emails when a DMCP user submits a question', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const dmcpUser: CMSUserType = {
        ...cmsUser,
        divisionAssignment: 'DMCP',
    }
    const questionFromDMCP: Question = {
        ...question,
        addedBy: dmcpUser,
    }
    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        questionFromDMCP
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            toAddresses: expect.arrayContaining([
                ...stateAnalysts,
                ...testEmailConfig().dmcpEmails,
            ]),
        })
    )
})

test('subject line is correct', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const name = packageName(
        sub.contract.stateCode,
        sub.contract.stateNumber,
        sub.formData.programIDs,
        defaultStatePrograms
    )

    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        question
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            subject: expect.stringContaining(`Questions sent for ${name}`),
            bodyText: expect.stringContaining(`${name}`),
        })
    )
})

test('includes link to the question response page', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        question
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringMatching(/View submission Q&A/),
            bodyHTML: expect.stringContaining(
                `http://localhost/submissions/${sub.contract.id}/question-and-answer`
            ),
        })
    )
})

test('includes expected data on the CMS analyst who sent the question', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs

    const template = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        question
    )

    if (template instanceof Error) {
        throw template
    }

    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining(
                'Sent by: Ronald McDonald (DMCO)  cms@email.com (cms@email.com)'
            ),
        })
    )
    expect(template).toEqual(
        expect.objectContaining({
            bodyText: expect.stringContaining('Date: 01/01/2024'),
        })
    )
})

test('renders overall email for a new question as expected', async () => {
    const sub = mockContractRev()
    const defaultStatePrograms = mockMNState().programs
    const result = await sendQuestionCMSEmail(
        sub,
        stateAnalysts,
        testEmailConfig(),
        defaultStatePrograms,
        question
    )

    if (result instanceof Error) {
        console.error(result)
        return
    }

    expect(result.bodyHTML).toMatchSnapshot()
})
