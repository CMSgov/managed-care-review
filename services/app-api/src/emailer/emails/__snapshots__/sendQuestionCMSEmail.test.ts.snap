// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders overall email for a new question as expected 1`] = `
"DMCO sent questions to the state for submission MCR-MN-0003-SNBC<br />
<b>Sent by:</b> Ronald McDonald (DMCO) <a href="cms@email.com">cms@email.com</a>
<br />
<b>Round:</b> 1<br />
<b>Date:</b> 12/31/2023<br />
<br />
<a href="http://localhost/submissions/12345/question-and-answers">View submission Q&A</a>
"
`;
